### PR TITLE
Support multi-target framework intellisense

### PR DESCRIPTION
**Pull request by mrward**
Type system service now supports multi-target frameworks.
Path bar supports multi-target framework selection.

MultiTargetFrameworkDropdown

MultiTargetIntellisense

Framework execution targets shown in main toolbar.

ExecutionTarget

SelectFrameworkExecutionTarget

Fixes VSTS #572311 - Multi-framework support in Roslyn workspace
Fixes VSTS #572342 - Support picking target framework for intellisense
Fixes VSTS #937105 Show (Multiple TargetFrameworks) in project options
Fixes VSTS #572309 - Execution target support for multi-framework projects
Fixes VSTS #935387 - System.NullReferenceException exception in MonoDevelop.Projects.DotNetProject.<OnGetReferencedAssemblies>d__XXX.MoveNext() - Transitive references null reference exception fixed since this code no longer existx.
Fixes VSTS #572307 - Ensure navigation/refactoring operations on multi-frameworks projects are coalesced
Fixes VSTS #843603 Files with build action 'None' are still used by Intellisense

 - [x] Cache framework specific configuration
     - Ensure the cache is refreshed when project options page changes config settings.
 - [x] OnExecute project extension should be framework aware
        - Do not want ASP.NET Core project extension checking for certificate status when .NET Framework is the execution target.
  - [x] Text editor path bar does not update after changing from a single framework to multiple.
    - [x] See type system errors after saving the edited .csproj - [Gist](https://gist.github.com/mrward/75870b499d8c111597f5b41cf1ad4b53) 
    - [x] Changing frameworks in .csproj does not always update the path bar. Sometimes see no owner projects, sometimes no workspace changed event is fired.
        - [x] After the project file reload the OpenDocumentReferences stored in the type system have an associated Project but its ParentSolution is null. This results in TryOpenDocumentInWorkspace doing nothing which presumably then breaks the editor path bar.

### CODE REVIEW

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ItemFrameworkConfigurationSelector.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		public override ItemConfiguration GetConfiguration (IConfigurationTarget target)
+		{
+			var original = configuration.GetConfiguration (target);
+
+			if (original is DotNetProjectConfiguration itemConfiguration) {
+				ItemConfiguration.ParseConfigurationId (original.Id, out string name, out string platform);
+				var copy = new DotNetProjectConfiguration (name, platform, framework);
+				copy.CopyFrom (itemConfiguration);
+
+				return copy;
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ItemFrameworkConfigurationSelector.cs]: That is weird. A selector is supposed to return one of the configurations in the provided target, not create a new one. Why is this required?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ItemConfiguration.cs [hunk 2]
```diff
@@ -84,17 +91,28 @@ public static void ParseConfigurationId (string id, out string name, out string
...
+					return name;
 			}
 		}
 		
 		public string Platform {
 			get { return platform ?? string.Empty; }
 		}
 
+		public string Framework {
+			get { return framework; }
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ItemConfiguration.cs]: Isn't the concept of framework is specific to .net? in that case this should be part of DotNetProjectConfiguration.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ItemFrameworkConfigurationSelector.cs (hunk 1 above)
Comment (mrward) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ItemFrameworkConfigurationSelector.cs]: Honestly this is a bit of a hack which I do not like.

The framework information is needed when running an MSBuild target so I was trying to get this to the DotNetProject when it runs MSBuild to get references etc without adding an extra framework parameter to each method. Currently the active selector in the workbench has no knowledge of the framework so the type system service is cheating here and creates another selector and sets the framework it wants.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp/CSharpPathedDocumentExtension.cs [hunk 3]
```diff
@@ -614,5 +618,32 @@ void CancelUpdatePath ()
 			src.Cancel ();
 			src = new CancellationTokenSource ();
 		}
+
+		class OwnerProjectInfo
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp/CSharpPathedDocumentExtension.cs]:
```suggestion
		sealed class OwnerProjectInfo
```

Comment (Therzok) [reply main/src/addins/CSharpBinding/MonoDevelop.CSharp/CSharpPathedDocumentExtension.cs]: Maybe, since we implemented Equals and GetHashCode, we should also:
```suggestion
		class OwnerProjectInfo : IEquatable<T>
```

File: main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.NodeBuilders/DependenciesNodeBuilder.cs [hunk 4]
```diff
@@ -62,22 +63,37 @@ public override bool HasChildNodes (ITreeBuilder builder, object dataObject)
...
+		public override void OnNodeRemoved (object dataObject)
+		{
+			var dependenciesNode = (DependenciesNode)dataObject;
+			dependenciesNode.PackageDependencyCache.PackageDependenciesChanged -= OnPackageDependenciesChanged;
+		}
+
+		void OnPackageDependenciesChanged (object sender, EventArgs e)
+		{
+			var cache = (PackageDependencyNodeCache)sender;
+			var project = cache.Project;
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.NodeBuilders/DependenciesNodeBuilder.cs]: Nit: cache.Project not used elsewhere, seems like a verbose local.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs [hunk 5]
```diff
@@ -1193,50 +1230,15 @@ protected override async Task OnClearCachedData ()
...
-			SetProperty (metadata, "ReferenceSourceTarget", reference.ReferenceSourceTarget);
-
-			return new AssemblyReference (path, metadata);
-		}
-
-		void SetProperty (MSBuildPropertyGroupEvaluated metadata, string name, string value)
-		{
-			var property = new MSBuildPropertyEvaluated (MSBuildProject, name, value, value);
-			metadata.SetProperty (name, property);
+			await base.OnClearCachedData ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs]: Not sure this method needs to be async anymore.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs [hunk 6]
```diff
@@ -1571,16 +1577,23 @@ static string[] GetTargetFrameworks (MSBuildProject project)
 			return null;
 		}
 
+		internal protected IEnumerable<string> GetTargetFrameworks ()
+		{
+			var frameworks = GetTargetFrameworks (MSBuildProject);
+			return frameworks ?? Enumerable.Empty<string> ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]:
```suggestion
			return frameworks ?? Array.Empty<string> ();
```

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs [hunk 7]
```diff
@@ -266,86 +267,108 @@ internal protected override async Task OnReevaluateProject (ProgressMonitor moni
...
-			ConfigurationSelector configuration,
-			System.Threading.CancellationToken token)
+		internal protected override string OnGetDefaultBuildAction (string fileName)
 		{
-			var references = new List<AssemblyReference> ();
-
-			var traversedProjects = new HashSet<string> (StringComparer.OrdinalIgnoreCase);
-			traversedProjects.Add (Project.ItemId);
+			string include = MSBuildProjectService.ToMSBuildPath (Project.ItemDirectory, fileName);
+			var globItems = Project.MSBuildProject.FindGlobItemsIncludingFile (include).ToList ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs]: Maybe we can do:
```suggestion
			var globItems = Project.MSBuildProject.FindGlobItemsIncludingFile (include).SingleOrDefault ();
```

To avoid matching all the items.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs [hunk 8]
```diff
@@ -266,86 +267,108 @@ internal protected override async Task OnReevaluateProject (ProgressMonitor moni
...
+			SolutionItemRunConfiguration runConfig)
 		{
-			foreach (var reference in await base.OnGetReferences (configuration, token).ConfigureAwait (false)) {
-				if (!reference.IsProjectReference) {
-					if (includeNonProjectReferences) {
-						references.Add (reference);
-					}
-					continue;
-				}
+			var targets = new List<ExecutionTarget> ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs]:
```suggestion
			var targets = new List<ExecutionTarget> (Project.TargetFrameworkMonikers.Length);
```

**Comment by Therzok** [comment]
LGTM!

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs (hunk 7 above)
Comment (mrward) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs]: Not sure I can change this. There may not be any matches - but could use FirstOrDefault instead. If there is more than one match we do not know what to return so fallback to asking the project for its default build action. Although not sure I have seen any examples of multiple matches here.

Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SdkProjectExtension.cs]: Yeah, was just trying to avoid a new collection allocation, as it seems redundant.

File: main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.Gui/DotNetCoreRuntimeOptionsPanelWidget.cs [hunk 9]
```diff
@@ -49,43 +49,49 @@ public DotNetCoreRuntimeOptionsPanelWidget (DotNetProject project)
...
-				if (cmp != 0)
-					return cmp;
-				return string.CompareOrdinal (y.Id.Version, x.Id.Version);
-			});
+				//sort by id ascending, version descending
+				frameworks.Sort ((x, y) => {
+					var cmp = string.CompareOrdinal (x.Id.Identifier, y.Id.Identifier);
+					if (cmp != 0)
+						return cmp;
+					return string.CompareOrdinal (y.Id.Version, x.Id.Version);
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.Gui/DotNetCoreRuntimeOptionsPanelWidget.cs]: I know this was already broken in previous versions and not a big cause of worry for now, but, for example, given the same identifier, a framework list with versions `8.0`, `9.0`, `10.0` would be sorted incorrectly, as:
```
9.0
8.0
10.0
```

Filed https://devdiv.visualstudio.com/DevDiv/_workitems/edit/950504

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/TargetFrameworkExecutionTarget.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+namespace MonoDevelop.Core.Execution
+{
+	internal class TargetFrameworkExecutionTarget : ExecutionTarget
+	{
+		public TargetFrameworkExecutionTarget (TargetFrameworkMoniker framework)
+		{
+			Id = framework.ShortName;
+			Name = framework.ShortName;
+			Framework = framework;
+			Image = "md-framework-dependency";
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/TargetFrameworkExecutionTarget.cs]: Should we have a constant for this in Stock?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs [hunk 11]
```diff
@@ -2087,6 +2088,131 @@ protected override async Task OnReevaluateProject (ProgressMonitor monitor)
...
+				return frameworkSpecificConfigurations.TryGetValue ((name, platform, framework), out configuration);
+		}
+
+		void CacheFrameworkSpecificConfiguration (string name, string platform, string framework, DotNetProjectConfiguration configuration)
+		{
+			lock (frameworkSpecificConfigurationsLock)
+				frameworkSpecificConfigurations [(name, platform, framework)] = configuration;
+		}
+
+		object frameworkSpecificConfigurationsLock = new object ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs]:
```suggestion
		readonly object frameworkSpecificConfigurationsLock = new object ();
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.OptionPanels/RuntimeOptionsPanel.cs [hunk 12]
```diff
@@ -72,46 +72,52 @@ public RuntimeOptionsPanelWidget (DotNetProject project, IEnumerable<ItemConfigu
...
+					cmp = string.CompareOrdinal (y.Id.Version, x.Id.Version);
+					if (cmp != 0)
+						return cmp;
+					return string.CompareOrdinal (x.Id.Profile, y.Id.Profile);
+				});
+
+				for (int i = 0; i < frameworks.Count; i++) {
+					var fx = frameworks[i];
+					if (project.TargetFramework.Id == fx.Id) {
+						string name = notInstalled? GettextCatalog.GetString ("{0} (Not installed)", fx.Name) : fx.Name;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.OptionPanels/RuntimeOptionsPanel.cs]:
```suggestion
						string name = notInstalled ? GettextCatalog.GetString ("{0} (Not installed)", fx.Name) : fx.Name;
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.MetadataReferenceHandler.cs [hunk 13]
```diff
@@ -65,57 +65,70 @@ ImmutableArray<MonoDevelopMetadataReference> CreateDefaultMetadataReferences ()
...
-				if (!(proj is MonoDevelop.Projects.DotNetProject netProject)) {
-					return CreateDefaultMetadataReferences ();
-				}
+				if (!string.IsNullOrEmpty (framework))
+					config = new MonoDevelop.Projects.DotNetProjectFrameworkConfigurationSelector (config, framework);
 
-				var data = new AddMetadataReferencesData {
-					Result = new List<MonoDevelopMetadataReference> (),
+				var data = new AddReferencesData {
+					References = new List<MonoDevelopMetadataReference> (),
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.MetadataReferenceHandler.cs]: Note to file a bug to use `ImmutableArray.Builder`

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectDataMap.cs [hunk 14]
```diff
@@ -48,56 +48,89 @@ public ProjectDataMap (MonoDevelopWorkspace workspace)
 				Workspace = workspace;
 			}
 
-			internal ProjectId GetId (MonoDevelop.Projects.Project p)
+			internal ProjectId GetId (MonoDevelop.Projects.Project p, string framework = null)
 			{
 				lock (gate) {
-					return projectIdMap.TryGetValue (p, out ProjectId result) ? result : null;
+					if (projectIdMap.TryGetValue (p, out var frameworkMappings)) {
+						var map = frameworkMappings.FirstOrDefault (f => f.Framework == framework);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectDataMap.cs]: Don't want to prematurely optimize, but `GetId` is called consistently a lot throughout the IDE usage.

Can we get rid of the lambda capture and LINQ usage?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectDataMap.cs [hunk 15]
```diff
@@ -164,6 +214,18 @@ internal ProjectId[] GetProjectIds ()
 					return projectDataMap.Keys.ToArray ();
 				}
 			}
+
+			class FrameworkMap
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectDataMap.cs]: Nit:
```suggestion
			sealed class FrameworkMap
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectSystemHandler.cs [hunk 16]
```diff
@@ -175,29 +199,45 @@ VersionStamp GetVersionStamp (MonoDevelop.Projects.Project project)
...
+				if (dotNetProject == null)
+					return null;
+
 				var workspace = Runtime.PeekService<RootWorkspace> ();
-				return workspace != null ? p.GetConfiguration (workspace.ActiveConfiguration) as MonoDevelop.Projects.DotNetProjectConfiguration : p.DefaultConfiguration as MonoDevelop.Projects.DotNetProjectConfiguration;
+				var config = workspace != null ? p.GetConfiguration (workspace.ActiveConfiguration) as DotNetProjectConfiguration : p.DefaultConfiguration as DotNetProjectConfiguration;
+				if (config == null || string.IsNullOrEmpty (framework))
+					return config;
+
+				return await dotNetProject.GetConfigurationAsync (config.Name, config.Platform, framework);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectSystemHandler.cs]:
```suggestion
				return await dotNetProject.GetConfigurationAsync (config.Name, config.Platform, framework).ConfigureAwait (false);
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectSystemHandler.cs [hunk 17]
```diff
@@ -97,11 +97,27 @@ class SolutionData
...
+				MonoDevelop.Projects.Project p,
+				CancellationToken token,
+				MonoDevelop.Projects.Project oldProject,
+				ProjectCacheInfo cacheInfo,
+				string framework)
+			{
+				var projectId = projectMap.GetOrCreateId (p, oldProject, framework);
 
-				var config = GetDotNetProjectConfiguration (p);
+				var config = await GetDotNetProjectConfiguration (p, framework);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectSystemHandler.cs]:
```suggestion
				var config = await GetDotNetProjectConfiguration (p, framework).ConfigureAwait (false);
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 18]
```diff
@@ -1497,24 +1498,26 @@ void OnProjectModified (object sender, MonoDevelop.Projects.SolutionItemModified
...
-							}
-							try {
-								lock (projectModifyLock) {
-									ProjectInfo newProjectContents = t.Result;
-									newProjectContents = AddVirtualDocuments (newProjectContents);
-									OnProjectReloaded (newProjectContents);
-									Runtime.RunInMainThread (() => IdeServices.TypeSystemService.UpdateRegisteredOpenDocuments ()).Ignore();
+						foreach (string framework in ProjectHandler.GetFrameworks (project)) {
+							var projectInfo = ProjectHandler.LoadProject (project, cts.Token, null, null, framework).ContinueWith (t => {
+								if (t.IsCanceled)
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Modify the ContinueWith call to harden and simplify the code a bit:

Specify `TaskScheduler` (`Task.Default` from the looks of it) for deterministic scheduling of the continuation.

Specify `TaskContinuationOptions.NotOnCanceled` and remove this IsCanceled check.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs [hunk 19]
```diff
@@ -511,14 +511,31 @@ async void OnSolutionItemAdded (object sender, MonoDevelop.Projects.SolutionItem
...
 					if (oldProject != null) {
-						projectInfo = ws.AddVirtualDocuments (projectInfo);
-						ws.OnProjectReloaded (projectInfo);
+						oldProjectIds = ws.GetProjectIds (oldProject).ToHashSet ();
 					}
-					else {
-						ws.OnProjectAdded (projectInfo);
+
+					foreach (string framework in ws.GetFrameworks (project)) {
+						var projectInfo = await ws.LoadProject (project, CancellationToken.None, oldProject, framework);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs]: Can we not do this work on parallel?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/WorkspaceFilesCache.cs [hunk 20]
```diff
@@ -78,34 +78,57 @@ void LoadCache (Solution sol)
...
-					TryDeleteFile (cacheFilePath);
-					continue;
 				}
+
+				if (cacheList != null)
+					cachedItems [projectFilePath] = cacheList;
+			}
+		}
+
+		internal IEnumerable<string> GetFrameworks (Project p)
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/WorkspaceFilesCache.cs]: Wonder if there is a way to deduplicate this into an extension method without polluting the internal API.

**Comment by Therzok** [comment]
Done with review pass 1. Looks really good!

File: main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.Gui/DotNetCoreRuntimeOptionsPanelWidget.cs (hunk 9 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore.Gui/DotNetCoreRuntimeOptionsPanelWidget.cs]: Marking as resolved, as I don't think it's reasonable to expect a fix in this PR.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.MetadataReferenceHandler.cs (hunk 13 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.MetadataReferenceHandler.cs]: Nothing to gain here, did some benchmarking.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs (hunk 18 above)
Comment (mrward) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: What should TaskScheduler be set to?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs (hunk 19 above)
Comment (mrward) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs]: Just wondering if there is a benefit to doing it in parallel. Each framework would be locked when accessing project information, such as getting reference assemblies for that framework.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs (hunk 18 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: TaskScheduler.Default. The code seems to want the continuation to run on the threadpool.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs (hunk 19 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs]: Hmmm, the only thing I'm thinking about here is that if we have 10 projects targeting 2 frameworks, we now do twice the work serially.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/TargetFrameworkExecutionTarget.cs (hunk 10 above)
Comment (mrward) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/TargetFrameworkExecutionTarget.cs]: This icon is currently in the NuGet addin. I guess it needs to be moved to Core since it is being used there.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/WorkspaceFilesCache.cs (hunk 20 above)
Comment (mrward) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/WorkspaceFilesCache.cs]: For now I have changed the WorkspaceFilesCache to use the method on the MonoDevelopWorkspace.ProjectSystemHandler.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/TargetFrameworkExecutionTarget.cs (hunk 10 above)
Comment (mrward) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/TargetFrameworkExecutionTarget.cs]: Making this a stock icon does not help here. Stock Icons are in MonoDevelop.Ide not MonoDevelop.Core so I cannot use it from here.

Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core.Execution/TargetFrameworkExecutionTarget.cs]: Gotcha! Thanks!

**Comment by Therzok** [comment]
LGTM iteration 2

**Comment by mrward** [comment]
Test failure is happening on master so not specific to this branch.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs [hunk 21]
```diff
@@ -2919,21 +2932,21 @@ ProjectConfiguration LoadConfiguration (ProgressMonitor monitor, ConfigData cgrp
 			return config;
 		}
 
-		MSBuildProjectInstance CreateProjectInstaceForConfiguration (string conf, string platform, bool onlyEvaluateProperties = true)
+		protected internal MSBuildProjectInstance CreateProjectInstanceForConfiguration (string conf, string platform, string framework = null, bool onlyEvaluateProperties = true)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]: Does it need to be protected? It could just be internal.

Comment (mrward) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]: Changed it to be internal not protected.

**Review by Therzok — APPROVED** [review]

**Comment by Therzok** [comment]
Took it for a spin, works properly from what I've tested! Nice work!

**Comment by sgmunn** [comment]
UI test for project system is from Nuget changes and is fixed in master.
.Net core tests are from mono bug. 

Neither failings are related as far as we know.

**Review by sgmunn — APPROVED** [review]
works good

**Review by slluis — APPROVED** [review]
